### PR TITLE
Use history API to clear hash to remove # sign

### PR DESCRIPTION
**Pull request by jasonnutter**
There are scenarios where an application uses server-side and client-side routing, and do not want the empty hash left over, as it will trigger client-side routing, which is not desired. Falling back to old method in case `replaceState` is not supported.

### CODE REVIEW

File: change/msal-2020-10-15-15-29-10-core-remove-hash-history.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Use history API to clear hash to remove # sign",
+  "packageName": "msal",
+  "email": "janutter@microsoft.com",
+  "dependentChangeType": "patch",
```
Comment (jasonnutter) [inline change/msal-2020-10-15-15-29-10-core-remove-hash-history.json:6]:
```suggestion
  "dependentChangeType": "none",
```

**Review by tnorling — APPROVED** [review]

File: lib/msal-core/test/UserAgentApplication.spec.ts [hunk 2]
```diff
@@ -1989,16 +1987,19 @@ describe("UserAgentApplication.ts Class", function () {
...
             const successHash = testHashesForState(TEST_LIBRARY_STATE).TEST_SUCCESS_ID_TOKEN_HASH + TEST_USER_STATE_NUM;
 
             window.location.href = loginStartPage;
             window.location.search = userQueryString;
 
             sinon.stub(window, "parent").returns(window);
-            sinon.stub(window.location, "href").returns(loginStartPage + successHash)
+            sinon.stub(window.location, "href").returns(loginStartPage + successHash);
+            sinon.stub(window.history, "replaceState").callsFake((data, title, url) => {
+                window.location.hash = "";
```
Comment (jasonnutter) [inline lib/msal-core/test/UserAgentApplication.spec.ts:2001]: Not really sure of a better way to handle jsdom not fully supporting the `history` APIs.

File: change/@azure-msal-browser-2020-10-16-15-05-51-core-remove-hash-history.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Use history API to clear hash for msal-browser",
+  "packageName": "@azure/msal-browser",
+  "email": "janutter@microsoft.com",
+  "dependentChangeType": "patch",
```
Comment (jasonnutter) [inline change/@azure-msal-browser-2020-10-16-15-05-51-core-remove-hash-history.json:6]:
```suggestion
  "dependentChangeType": "none",
```

File: samples/msal-core-samples/VanillaJSTestApp/app/get_access_token_interactively/test/browser.spec.ts [hunk 4]
```diff
@@ -174,7 +174,7 @@ describe("Browser tests", function () {
         await takeScreenshot(page, testName, "samplePageGotToken");
         localStorage = await page.evaluate(() =>  Object.assign({}, window.localStorage));
         expect(Object.keys(localStorage).length).to.be.eq(6);
-        expect(page.url()).to.be.eq(customQueryPage + "#"); // Redirect will leave empty hash on the url
+        expect(page.url()).to.be.eq(customQueryPage); // Redirect will leave empty hash on the url
```
Comment (jasonnutter) [inline samples/msal-core-samples/VanillaJSTestApp/app/get_access_token_interactively/test/browser.spec.ts:177]:
```suggestion
        expect(page.url()).to.be.eq(customQueryPage);
```

**Review by pkanher617 — APPROVED** [review]

File: lib/msal-browser/src/utils/BrowserUtils.ts [hunk 5]
```diff
@@ -32,7 +32,12 @@ export class BrowserUtils {
      * Clears hash from window url.
      */
     static clearHash(): void {
-        window.location.hash = "";
+        if ("replaceState" in history) {
```
Comment (pkanher617) [inline lib/msal-browser/src/utils/BrowserUtils.ts:35]: Should this be done in 1.x?

Comment (pkanher617) [reply lib/msal-browser/src/utils/BrowserUtils.ts:35]: Sorry nvm.

**Pull request merged**